### ISSUE TITLE
Make asset handles `!Default`

### ISSUE THREAD

**Issue by RedlineTriad** [bevyengine/bevy#2647]
## What problem does this solve or, what need does it fill?

[Handle](https://github.com/bevyengine/bevy/blob/b13472dae495ecc7d1a80cbe1ad12718c0de9262/crates/bevy_asset/src/handle.rs#L61) currently requires `Default` to be implemented, which reduces type safety because for many `Assets` no default makes sense. Removing this requirement would also partially solve #1201, because when creating `Text` and filling out the values with `..Default::default()` `style.font` will be set to an invalid font Handle and fail to render.

## What solution would you like?

The removal of the requirement that all `Handles` have to implement `Default`. The reason this requirement exists is because `Handle` derives `Reflect`.

## What alternative(s) have you considered?

Runtime errors for default handle values where `Default` is invalid, but this delays the problem feedback and increases iteration time compared to getting the error at compile time.

**Comment by alice-i-cecile** [comment]
#1395 is related, covering a `Default`-like trait specifically designed for structs created via reflection (such as objects loaded from scenes).

**Comment by NiklasEi** [comment]
I agree that the default impl for all handles reduces type safety. I am wondering though whether removing it would reduce the usibility of many build-in bundles (at least if that also means the bundles themself no longer are `Default`).

If there is no longer a default for the font handle in the text bundle, what does the bundle set as default? In the case of fonts this might be solved with ongoing discussions about bundling fonts/using system fonts (#1017, #1325), but the same problem arises with other bundles. E.g. is there a default for sprite bundle if there is no default for texture handles?

One idea: we could have convenience functions like `with_font` or `with_texture` that still allow using default values for all other fields. But this is still not as nice to use as the current `..Default::default()`.

**Comment by RedlineTriad** [comment]
> I agree that the default impl for all handles reduces type safety. I am wondering though whether removing it would reduce the usibility of many build-in bundles (at least if that also means the bundles themself no longer are `Default`).

It is only the *requirement* I care about. If a default font is added that can be used, and `Default` implemented. But it shouldn't be required to be implemented for all assets, even those where no such reasonable default exists.

**Comment by NiklasEi** [comment]
Only lifting the requirement would not improve your example of the Text bundle. We should then also rethink defaults for all bundles.

But just lifting the requirement sounds reasonable as a first step 👍

**Comment by RedlineTriad** [comment]
> Only lifting the requirement would not improve your example of the Text bundle. We should then also rethink defaults for all bundles.

Sorry, what I wrote was a bit misleading. I do care about removing the Default implementations, but only where they don't currently make sense,
such as `Text`. 
So I want `Default` to be optional and implemented where a reasonable default exists; and not where it doesn't.

**Comment by mockersf** [comment]
While I would like to remove the `Default` impl for `Handle<T>` as it indeed doesn't make sense, it would break what is currently the favoured api (using `..Default::default()`).

It could be useful to list all cases where a bundle has an handle component, and see what default could be used there
* `Handle<Texture>` -> Bevy logo?
* `Handle<Mesh>` -> a cube?
* ...

**Comment by RedlineTriad** [comment]
>     * `Handle<Texture>` -> Bevy logo?

It is common to have some kind of very obvious missing texture, like fully saturated purple, maybe even checkered to make it even more obvious.
![image](https://user-images.githubusercontent.com/39059512/129846953-96368ee8-603b-40dd-a473-146ac4719e33.png)
